dominion/src/main/java/org/codice/dominion/interpolate/Interpolator.java

```diff
@@ -0,0 +1,436 @@
+          new Gson().fromJson(properties.getProperty(Interpolator.REPLACEMENTS_KEY, ""), Map.class);
+    } catch (JsonSyntaxException e) {
+      throw new InterpolationException("Unable to determined replacement information", e);
+    }
+    this.id = initFromReplacements("test.id");
+    this.container = initFromReplacements("container.name");
+    try {
+      final PortFinder portFinder =
+          new Gson().fromJson(properties.getProperty(Interpolator.PORTS_KEY, ""), PortFinder.class);
+
+      this.ports =
+          (portFinder != null)
+              ? portFinder
+              : new PortFinder(Interpolator.BASE_PORT, Interpolator.BLOCK_SIZE);
+    } catch (JsonSyntaxException e) {
+      throw new InterpolationException("Unable to determined reserved ports information", e);
+    }
+  }
+
+  /**
+   * Gets the unique identifier for the current test run.
+   *
+   * @return a unique identifier for the current test run
+   */
+  public String getUUID() {
+    return id;
+  }
+
+  /**
+   * Gets the name of the container associated with this interpolator.
+   *
+   * @return the container name
+   */
+  public String getContainer() {
+    return container;
+  }
+
+  /**
+   * Interpolates a given annotation by returning a proxy version that will automatically
+   * interpolate all <code>String</code> and <code>String[]</code> methods that are annotated with
+   * the {@link Interpolate} annotation.
+   *
+   * <p><<i>Note:</i> This method will not attempt any interpolations. Instead it will return a
+   * proxy which will delay interpolation until an annotation value marked to be interpolated is
+   * accessed. It is therefore possible that accessing a value of an interpolated annotation throws
+   * an instance of {@link InterpolationException}.
+   *
+   * @param <A> the type of annotation to be interpolated
+   * @param annotation the annotation to be interpolated
+   * @return <code>annotation</code> if no methods are requested to be interpolated or a proxy if
+   *     interpolation was requested
+   */
+  public <A extends Annotation> A interpolate(A annotation) {
+    // check if the annotation has methods annotated with @Interpolate in which case, we will want
+    // to create a proxy for the annotation such that we can report interpolated values instead
+    final Class<? extends Annotation> clazz = annotation.annotationType();
+    final Set<Method> toInterpolate =
+        methodsToInterpolate.computeIfAbsent(
+            clazz,
+            c ->
+                Stream.of(c.getMethods())
+                    .filter(m -> m.getAnnotation(Interpolate.class) != null)
+                    .collect(Collectors.toSet()));
+
+    if (toInterpolate.isEmpty()) {
+      return annotation;
+    }
+    return (A)
+        Proxy.newProxyInstance(
+            annotation.getClass().getClassLoader(),
+            new Class<?>[] {annotation.annotationType(), Annotation.class},
+            new InvocationHandler() {
```

Comment:
✏️ I think moving the creation of this InvocationHandler outside of this method would make this a little easier to read.